tests/test_tpp_backend.py

```diff
@@ -2851,3 +2851,63 @@ def test_extracting_at_different_index_dates():
     study.set_index_date("2020-02-01")
     results = study.to_dicts()
     assert_results(results, value=["", "2020-02-10"])
+
+
+def test_aggregate_over_different_date_formats_raises_error():
+    with pytest.raises(ValueError, match="dates of different format"):
+        StudyDefinition(
+            population=patients.all(),
+            birth=patients.date_of_birth(date_format="YYYY-MM"),
+            death=patients.with_death_recorded_in_primary_care(
+                returning="date_of_death", date_format="YYYY-MM-DD"
+            ),
+            last_major_life_event=patients.maximum_of("birth", "death"),
+        )
+
+
+def test_dynamic_index_dates():
+    session = make_session()
+    session.add_all(
+        [
+            Patient(
+                CodedEvents=[
+                    CodedEvent(ConsultationDate="2020-01-01", CTV3Code="foo"),
+                    CodedEvent(ConsultationDate="2020-02-01", CTV3Code="foo"),
+                    CodedEvent(ConsultationDate="2020-03-01", CTV3Code="bar"),
+                    CodedEvent(ConsultationDate="2020-04-20", CTV3Code="foo"),
+                    CodedEvent(ConsultationDate="2020-05-01", CTV3Code="foo"),
+                    CodedEvent(ConsultationDate="2020-06-01", CTV3Code="bar"),
+                ],
+            ),
+        ]
+    )
+    session.commit()
+    study = StudyDefinition(
+        population=patients.all(),
+        earliest_bar=patients.with_these_clinical_events(
+            codelist(["bar"], system="ctv3"),
+            returning="date",
+            date_format="YYYY-MM-DD",
+            find_first_match_in_period=True,
+        ),
+        latest_foo_before_bar=patients.with_these_clinical_events(
+            codelist(["foo"], system="ctv3"),
+            returning="date",
+            date_format="YYYY-MM-DD",
+            find_last_match_in_period=True,
+            on_or_before="earliest_bar",
+        ),
+        latest_foo_in_month_after_bar=patients.with_these_clinical_events(
+            codelist(["foo"], system="ctv3"),
+            returning="date",
+            date_format="YYYY-MM-DD",
+            find_last_match_in_period=True,
+            on_or_before="last_day_of_month(earliest_bar) + 1 month",
+        ),
+    )
+    assert_results(
+        study.to_dicts(),
+        earliest_bar=["2020-03-01"],
+        latest_foo_before_bar=["2020-02-01"],
+        latest_foo_in_month_after_bar=["2020-04-20"],
+    )
```

Comment:
It would be good to have a test that demonstrates the behaviour when a date expression is invalid